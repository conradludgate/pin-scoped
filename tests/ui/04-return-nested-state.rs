#![feature(async_closure)]

use std::pin::pin;

use pin_scoped::{Scope, ScopeState};

#[tokio::main]
async fn main() {
    let scoped = pin!(Scope::new(0));

    let handle = scoped
        .as_ref()
        .spawn(async move |state: ScopeState<u64>| state.nest(1));

    let mut state = scoped.await;
    let nested_scope = handle.await.unwrap();

    state += 1;
    println!("{state}");

    drop(nested_scope);
}
