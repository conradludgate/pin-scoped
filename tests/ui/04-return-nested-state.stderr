error: lifetime may not live long enough
  --> tests/ui/04-return-nested-state.rs:13:52
   |
13 |         .spawn(async move |state: ScopeState<u64>| state.nest(1));
   |                            -----                 - ^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |                            |                     |
   |                            |                     return type of async closure `{async closure body@$DIR/tests/ui/04-return-nested-state.rs:13:52: 13:65}` contains a lifetime `'2`
   |                            has type `ScopeState<'1, u64>`
   |
   = note: requirement occurs because of the type `NestedScope<'_, u64, i32>`, which makes the generic argument `'_` invariant
   = note: the struct `NestedScope<'parent, Head, State>` is invariant over the parameter `'parent`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
